src/Tokenomia/Vesting/PrivateSale.hs

```diff
@@ -0,0 +1,223 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Tokenomia.Vesting.PrivateSale (verifyPrivateSale) where
+
+import Ledger hiding (mint, singleton)
+import Ledger.Value as Value
+import Prelude hiding (readFile)
+
+import Control.Monad hiding (fmap)
+import Control.Monad.Reader hiding (ask)
+import GHC.Generics (Generic)
+
+import Debug.Trace
+
+import Data.Hex (unhex)
+import Data.String (fromString)
+
+import Control.Lens
+import Control.Monad.Except
+import Data.Aeson (FromJSON, decode, eitherDecodeFileStrict)
+import Data.Bifunctor (first)
+import Data.ByteString.Lazy (readFile)
+import Data.Foldable (find)
+import Data.Text (Text, pack, unpack)
+import Money
+import System.IO (FilePath, getLine, putStrLn)
+
+import qualified Blockfrost.Client as B
+import qualified Blockfrost.Lens   as B
+
+import Tokenomia.Common.Blockfrost (projectFromEnv'')
+import Tokenomia.Common.Environment
+import Tokenomia.Common.Error
+
+data PrivateSale = PrivateSale
+    { _psAddress :: B.Address -- Treasury address
+    , _psStart :: POSIXTime
+    , _psTranches :: [Tranche]
+    , _psAssetClass :: Value.AssetClass -- Vesting token
+    , _psInvestors :: [PrivateInvestor]
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+data Tranche = Tranche
+    { _tranchePercentage :: Integer
+    , _trancheDuration :: Integer
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+data PrivateInvestor = PrivateInvestor
+    { _piAddress :: B.Address
+    , _piAllocation :: Integer -- Amount of vesting tokens to lock (note, this is likely NOT equal to the total amounts of investments, as different tokens)
+    , _piInvestments :: [Investment]
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+data Investment = Investment
+    { _invTx :: B.TxHash -- Tx that sends below asset class to payment address
+    , _invAssetClass :: Value.AssetClass
+    , _invAmount :: Integer -- Amount of above asset class expected to be sent to payment address
+    }
+    deriving stock (Generic, Show)
+    deriving anyclass (FromJSON)
+
+makeLenses ''PrivateSale
+makeLenses ''Tranche
+makeLenses ''PrivateInvestor
+makeLenses ''Investment
+
+verifyPrivateSale ::
```

Comment:
This isn't something that Tokenomia is doing, but we're trying to follow our own styleguide so - we need explicit `foralls` and kind signatures.
So, in this case, `forall (m :: Type -> Type). `  prefixed.
https://github.com/mlabs-haskell/styleguide#type-and-kind-signatures